naptime/src/main/scala/org/coursera/naptime/ari/engine/EngineImpl.scala

```diff
@@ -2,42 +2,108 @@ package org.coursera.naptime.ari.engine
+import org.coursera.naptime.ari.TopLevelRequest
 import org.coursera.naptime.router2.NaptimeRoutes
 import org.coursera.naptime.schema.Resource
+import play.api.libs.json.JsString
 
+import scala.collection.JavaConverters._
 import scala.concurrent.ExecutionContext
 import scala.concurrent.Future
 
 class EngineImpl @Inject() (
     naptimeRoutes: NaptimeRoutes,
     fetcher: FetcherApi)
-    (implicit executionContext: ExecutionContext) extends EngineApi {
+    (implicit executionContext: ExecutionContext) extends EngineApi with StrictLogging {
 
-  override val schemas: Seq[Resource] = naptimeRoutes.routerBuilders.map(_.schema).toList
+  private[this] val resourceSchemas: Map[ResourceName, Resource] = naptimeRoutes.schemaMap.flatMap {
+    case (_, schema) if schema.parentClass.isEmpty => // TODO: handle sub resources
+      val resourceName = ResourceName(schema.name, version = schema.version.getOrElse(0L).toInt)
+      Some(resourceName -> schema)
+    case (_, schema) =>
+      logger.warn(s"Cannot handle nested resource $schema")
+      None
+  }
 
-  override val models: Map[String, DataSchema] =
-    naptimeRoutes.routerBuilders.foldLeft(Map[String, DataSchema]()) { case (allTypes, routerBuilder) =>
-      allTypes ++ routerBuilder.types.flatMap { keyedSchema =>
-        keyedSchema.value match {
-          case recordSchemaValue: DataSchema => Some(keyedSchema.key -> recordSchemaValue)
-          case _ => None
-        }
-      }
-    }
+  private[this] val mergedTypes = naptimeRoutes.routerBuilders.flatMap(_.types.map(_.tuple))
+    .filter(_._2.isInstanceOf[RecordDataSchema])
+    .map(tuple => tuple._1 -> tuple._2.asInstanceOf[RecordDataSchema]).toMap
 
   override def execute(request: Request): Future[Response] = {
     val responseFutures = request.topLevelRequests.map { topLevelRequest =>
-      val singleRequest = Request(request.requestHeader, List(topLevelRequest))
-      fetcher.data(singleRequest)
+      executeTopLevelRequest(request, topLevelRequest)
     }
     val futureResponses = Future.sequence(responseFutures)
     futureResponses.map { responses =>
       responses.foldLeft(Response.empty)(_ ++ _)
     }
   }
+
+  private[this] def executeTopLevelRequest(request: Request, topLevelRequest: TopLevelRequest): Future[Response] = {
+    val singleRequest = Request(request.requestHeader, List(topLevelRequest))
+    val topLevelResponse = fetcher.data(singleRequest)
+
+    topLevelResponse.flatMap { topLevelResponse =>
+      // If we have a schema, we can perform automatic resource inclusion.
+      (for {
+        topLevelSchema <- resourceSchemas.get(topLevelRequest.resource)
+        mergedType <- mergedTypes.get(topLevelSchema.mergedType)
+      } yield {
+        val topLevelData = topLevelResponse.data.get(topLevelRequest.resource).toIterable.flatMap { dataMap =>
+          dataMap.values
+        }
+        val nestedLookups = topLevelRequest.selection.selections.filter(_.selections.nonEmpty)
```

Comment:
Ah, you're quite right. I previously forgot about that. I think instead of throwing on line 74, I should handle it appropriately. I'll add a test and then the fix in a subsequent PR.